File: subArraySumEqualsK.java
import java.util.HashMap;

public class subArraySumEqualsK {

    // Time Complexity : O(N)
    // Space Complexity : O(N)
    // Did this code successfully run on Leetcode : Yes
    // Any problem you faced while coding this :No

    // Your code here along with comments explaining your approach
    // using the running sum pattern -> maintaining the cumulative sum and its
    // occurence

    public static void main(String[] args) {
        int count = 0, sum = 0;
        int nums[] = { 1, 1, 1 };
        int target = 2;
        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            if (map.containsKey(sum - target))
                count += map.get(sum - target);
            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }
        System.out.println(count);
    }
}

File: SubarraySumEqualsK.js
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
  if (nums.length === 0 || nums === null) {
    return -1;
  }

  var rSum = 0;
  var count = 0;
  var hMap = new Map();
  var rSum = 0;

  hMap.set(0, 1);

  for (var i = 0; i < nums.length; i++) {
    var n = Number(nums[i]);
    rSum = rSum + n;
    var y = rSum - k;
    if (hMap.has(y)) {
      count = count + Number(hMap.get(y));
    }

    if (hMap.has(rSum)) {
      var prevVal = hMap.get(rSum);
      hMap.set(rSum, prevVal + 1);
    } else {
      hMap.set(rSum, 1);
    }
  }
  return count;
};
